docker/main/docker-build.yml

```diff
@@ -49,6 +49,7 @@ services:
       context: ./ngen
       target: rocky-ngen-deps
       args:
+        NUMPY_CONSTRAINT: ${NUMPY_INSTALL_CONSTRAINT:?Please set global NUMPY_INSTALL_CONSTRAINT in environment config}
```

Comment:
This is an abuse of the `args` parameter. Do not use this to inject python versions. Stick to `setup.py` and `requirements.txt`.

Reply:
We don't exactly have a requirements.txt for _inside_ the worker images (they kind of use the one that's part of the ngen Python BMI tests, but that's not really what that is for).  I suppose we could go that way, but this then would have to become a subtle overhaul of what we are doing inside the images to install numpy, rather than an extension to what's already in place.  I.e., the scope, size, and impact would change significantly.  At that point, I think we'd have to at least reconsider whether to just wait for the numpy 2.0.0 compatibility issue to be fixed on the ngen side.

Aside from that, this also would have the side effect of committing to a particular numpy version in the code, rather than allowing users to configure that per deployment.  I'd be interested in hearing an argument as to whether that in and of itself is a preferable approach, if you're suggesting that.

Reply:
> I'd say it might be up for debate for us to really set a version - the installation of libraries like pandas will do that for us.

Eh, there isn't a way around us doing this, at least implicitly, because ngen needs to be compiled with some version of numpy before anything like that is installed.

> Relying on args in this context is an indication that we are doing something wrong. If this is the only solution, we need to take a step back and fix the mistakes that require this.

It isn't the only way to fix this problem, but it's the only way I see to fix it by providing dynamic user control over the numpy version, and there at least an argument for needing that when we dynamically inject user-provided BMI modules into job workers.

> Additionally, if ngen is broken, it is not DMOD's responsibility to fix it.

There's a debate there too.  ngen still works fine as long as you don't install numpy 2.0.  It's the DMOD images that are doing that.